File: PO_Projekt2/ProgOb_Projekt2/src/main/java/pl/Dybowski/Adrian/Projekt2/WirtualnySwiat/Rosliny/Roslina.java
package pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Rosliny;

import java.util.HashSet;

import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Organizm;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.RodzajeKomunikatow;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Swiat;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Wspolrzedne2D;

public abstract class Roslina extends Organizm {
	protected final int probyRozprzestrzeniania;
	protected final double szansaRozprzestrzeniania;
	
	public Roslina(Swiat swiat, Wspolrzedne2D polozenie, int sila, int inicjatywa,
					double szansaRozprzestrzeniania, int probyRozprzestrzeniania) {
		super(swiat, polozenie, inicjatywa, sila);
		this.probyRozprzestrzeniania = probyRozprzestrzeniania;
		this.szansaRozprzestrzeniania = szansaRozprzestrzeniania;
	}
	
	@Override
	public void akcja() {
		if (!zyje) return;
		
		HashSet<Wspolrzedne2D> pola = swiat.sasiednieWolnePola(polozenie);
		if (pola.size() == 0) return;

		for (int i = 0; i < probyRozprzestrzeniania; i++) {
			if (!mozeSieRozprzestrzeniac()) continue;
			
			Organizm dziecko = stworzDziecko(swiat.losowePole(pola));
			swiat.dodajDziecko(dziecko);
			
			swiat.dodajKomunikat(dziecko, RodzajeKomunikatow.wyrasta);
		}
	}
	
	@Override
	public void kolizja(Organizm ofiara, boolean atak) {
		if (atak) return;
		
		if (zyje) zakonczZycie();
	}
	
	private boolean mozeSieRozprzestrzeniac() {
		return swiat.szansaTrafiona(szansaRozprzestrzeniania);
	}
}

File: projekty/PO_Projekt3/roslina.py
from organizm import Organizm
import random

class Roslina(Organizm):
    def __init__(self, swiat, polozenie, sila, inicjatywa, szansaRozprzestrzeniania, probyRozprzestrzeniania):
        super().__init__(swiat, polozenie, sila, inicjatywa)
        self._szansaRozprzestrzeniania = szansaRozprzestrzeniania
        self._probyRozprzestrzeniania = probyRozprzestrzeniania

    def get_szansaRozprzestrzeniania(self):
        return self._szansaRozprzestrzeniania

    def set_szansaRozprzestrzeniania(self, szansaRozprzestrzeniania):
        self._szansaRozprzestrzeniania = szansaRozprzestrzeniania

    def get_probyRozprzestrzeniania(self):
        return self._probyRozprzestrzeniania

    def set_probyRozprzestrzeniania(self, probyRozprzestrzeniania):
        self._probyRozprzestrzeniania = probyRozprzestrzeniania

    def akcja(self):
        if not self._zyje:
            return
        
        pola = self._swiat.sasiednieWolnePola(self._polozenie)
        
        if len(pola) == 0:
            return
        
        for _ in range(self._probyRozprzestrzeniania):
            if not self.mozeSieRozprzestrzeniac():
                continue

            dziecko = self.stworzDziecko(self._swiat.losowePole(pola))
            self._swiat.dodajDziecko(dziecko)
            self._swiat.dodajKomunikat(str(self._polozenie) + ": " + self.nazwaKlasy() + " rozprzestrzenia sie")

    def kolizja(self, ofiara, atak):
        if atak:
            return

        if self._zyje:
            self.zakonczZycie()

    def mozeSieRozprzestrzeniac(self):
        return random.random() < self._szansaRozprzestrzeniania




